### PR TITLE
vmselect: update `vm_request_duration_seconds` value when request fails

### PR DESCRIPTION
**Pull request by hagen1778**
Before, metric `vm_request_duration_seconds` was update only on successful
attempts which could be misleading. For example, timeout errors on netstorage
request may be not accounted in the metric and won't be visible on dashboards.
Using `defer` statement to update the metric after query arguments validation
may improve the situation.

### CODE REVIEW

File: app/vmselect/prometheus/prometheus.go [hunk 1]
```diff
@@ -76,6 +76,7 @@ func FederateHandler(startTime time.Time, w http.ResponseWriter, r *http.Request
 	if err != nil {
 		return err
 	}
+	defer federateDuration.UpdateDuration(startTime)
```
Comment (valyala) [inline app/vmselect/prometheus/prometheus.go:79]: I think it is better from readability and maintainability PoV moving these `defer` calls to the first line of every measured function.

Comment (hagen1778) [reply app/vmselect/prometheus/prometheus.go:79]: This will account for queries which fail the validation. Do you think it matters?

Comment (valyala) [reply app/vmselect/prometheus/prometheus.go:79]: I think this is OK

Comment (hagen1778) [reply app/vmselect/prometheus/prometheus.go]: updated

**Review by valyala — APPROVED** [review]
LGTM

**Comment by valyala** [comment]
FYI, this pull request has been included in [v1.65.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by shirongxiao** [comment]
Using defer statement to update the metric after query
When I execute a query, the vmselect log shows:
error in "/select/0/prometheus/api/v1/query_range?..": error when executing query="...": cannot evaluate "...": error occurred during search: cannot fetch query results from vmstorage nodes: cannot perform search on vmstorage: request timeout reached: 0.010 seconds (elapsed 0.643 seconds); the timeout can be adjusted with \-[search.ma](https://search.ma/)xQueryDuration` command-line flag`

I am confused about why the elapsed value in the log differs from the value of the vm_request_duration_seconds metric. The value of vm_request_duration_seconds is much smaller than the elapsed value, such as 100ms. @hagen1778 @valyala

**Comment by makasim** [comment]
> I am confused about why the elapsed value in the log differs from the value of the vm_request_duration_seconds metric. > The value of vm_request_duration_seconds is much smaller than the elapsed value, such as 100ms. 

The question was answered in https://github.com/VictoriaMetrics/VictoriaMetrics/issues/9689#issuecomment-3267411149

**Pull request merged**